openedx/core/djangoapps/oauth_dispatch/models.py

```diff
@@ -43,3 +49,76 @@ def verify_access_token_as_expired(cls, access_token):
         is set at the beginning of the epoch which is Jan. 1, 1970
         """
         return access_token.expires == datetime(1970, 1, 1, tzinfo=utc)
+
+
+class ScopedApplication(AbstractApplication):
+    """
+    Custom Django OAuth Toolkit Application model that enables the definition
+    of scopes that are authorized for the given Application.
+    """
+    FILTER_USER_ME = 'user:me'
+ 
+    scopes = ListCharField(
+        base_field=models.CharField(max_length=32),
+        size=10,
+        max_length=(10 * 33),  # 10 * 32 character scopes, plus commas
+        help_text=_('Comma-separated list of scopes that this application will be allowed to request.'),
+    )
+ 
+    def __unicode__(self):
+        """
+        Return a unicode representation of this object.
+        """
+        return u"<ScopedApplication '{name}'>".format(
+            name=self.name
+        )
+ 
+    @property
+    def authorization_filters(self):
+        """
+        Return the list of authorization filters for this application.
+        """
+        filters = [':'.join([org.provider_type, org.short_name]) for org in self.organizations.all()]
+        if self.authorization_grant_type == self.GRANT_CLIENT_CREDENTIALS:
+            filters.append(self.FILTER_USER_ME)
+        return filters
+ 
+ 
+class ScopedApplicationOrganization(models.Model):
+    """
+    Associates an organization to a given ScopedApplication including the
+    provider type of the organization so that organization-based filters
+    can be added to access tokens provided to the given Application.
+ 
+    See openedx/core/djangoapps/oauth_dispatch/docs/decisions/0007-include-organizations-in-tokens.rst
+    for the intended use of this model.
+    """
+    CONTENT_PROVIDER_TYPE = 'content_org'
+    ORGANIZATION_PROVIDER_TYPES = (
+        (CONTENT_PROVIDER_TYPE, _('Content Provider')),
+    )
+ 
+    # In practice, short_name should match the short_name of an Organization model.
+    # This is not a foreign key because the organizations app is not installed by default.
+    short_name = models.CharField(
+        max_length=255,
+        help_text=_('The short_name of an existing Organization.'),
+    )
+    provider_type = models.CharField(
+        max_length=32,
+        choices=ORGANIZATION_PROVIDER_TYPES,
+        default=CONTENT_PROVIDER_TYPE,
+    )
+    application = models.ForeignKey(
+        oauth2_settings.APPLICATION_MODEL,
```

Comment:
Should this be dependent on the setting - or always related to `ScopedApplication` since they seem to be sister models?  Each way has its own trade-offs, of course.  I suppose the current way is more extensible, right?  

What if someone changes the setting?  That won't result in a new migration.  But that should be Ok, I believe.

Reply:
Yes, I think this is good as is.